### ISSUE TITLE
Refreshing ads on single page applications causes excessive memory usage. 

### ISSUE THREAD

**Issue by 12q1** [prebid/Prebid.js#10880]
## Type of issue
Question

## Description
When creatives are given access to the top window object they insert third party scripts and create additional DOM nodes. On traditional static HTML pages this is almost unpercievable but when combined with ad refreshes on single page applications these references are not cleaned up and can heavily impact performance. 

## Steps to reproduce
Can be simulated on single page applications by navigating from page to page or waiting for multiple refreshes. 

## Test page
Issue can be observed on: https://www.puna.nl

### Expected results
Expect creatives to be unloaded gracefully after refreshing and remove references/DOM nodes from the page.  

### Actual results
Over time the memory (Read Total JS Heap) increases throughout the user journey. In worst case scenarios the memory usage can become so high that it locks up the client's browser.

![image](https://github.com/prebid/Prebid.js/assets/42697100/12bb2f13-593a-45ee-a1c9-418f6bda827a)

## Platform details
All web integrations regardless of browser/OS/build version. 

## Other information
Seems to be a known issue:
- https://github.com/prebid/Prebid.js/issues/352
- https://github.com/prebid/Prebid.js/issues/1328

We are curious if there has been any changes to Prebid's recommendation to force safe frames to prevent creatives from breaking out of their sandbox and accessing the top window. Some of our rich media partners prefer to work without safe frames so this creates a problem when trying to support their integrations. 

Thanks.

**Comment by patmmccann** [comment]
are you taking advantage of #10308

**Comment by 12q1** [comment]
> are you taking advantage of #10308

@patmmccann yes, our prebid config object currently contains the following settings:

```
pbjs.setConfig({
        minBidCacheTTL: 30,
        eventHistoryTTL: 30,
});
```

Would you suggest putting these options even lower? We did see some minor improvement from these setting but the bulk of the memory usage still seems to be caused by creatives leaving references (eg. event listeners) and DOM nodes on the page.

**Comment by patmmccann** [comment]
Are you seeing an issue where safari mobile just refreshes pages?

**Comment by patmmccann** [comment]
Monkey patching insertelement, add listener etc, you could tag the element with a creative id and clean them up later. This is quite onerous. It seems a community push towards safeframe is likely a better solution. We're likely able to solve some cleanup issues with very specific examples. Maybe let's open more issues of this nature detailing some of the remnant items

**Comment by 12q1** [comment]
@patmmccann thanks for the update, can't say we've noticed the safari issue but will take a look.

Regarding pushing for safeframes I couldn't agree more but as it stands our partners are not willing to relinquish their access to the top window. Have to balance the business requirements against the risk here.

**Comment by jpm** [comment]
> Are you seeing an issue where safari mobile just refreshes pages?

We have noticed the page refresh issue both on Safari/iOS as well as Chrome/iOS. We are assuming that this is because both browsers use WebKit, the same underlying web browser engine.

It's been a difficult issue to troubleshoot, as Apple doesn't seem interested in dealing with the problem, and other vendors don't have a good way to avoid it.

**Comment by alnagar** [comment]
I am using a third party provider to serve me ads and they happen to use prebidjs. My users spend hours on my single page application (angular app) and the browser on phones are getting refreshed forcibly mid-conversation causing to cut the line. That's a terrible user experience. I checked the heap profiler in chrome on my desktop and the memory consumption go up rapidly. Does anyone look at this ? Is there any workraround I can implement on my end ? This bug kills the user experience :(

**Comment by patmmccann** [comment]
>  Is there any workraround I can implement on my end ? 

Safeframes are quite capable of making sure ads are destroyed more gracefully, unless of course, a renderer is provided and the safeframe is bypassed.

**Comment by 12q1** [comment]
> I am using a third party provider to serve me ads and they happen to use prebidjs. My users spend hours on my single page application (angular app) and the browser on phones are getting refreshed forcibly mid-conversation causing to cut the line. That's a terrible user experience. I checked the heap profiler in chrome on my desktop and the memory consumption go up rapidly. Does anyone look at this ? Is there any workraround I can implement on my end ? This bug kills the user experience :(

The only full-proof solution we could find was to enable [setForceSafeFrame ](https://developers.google.com/publisher-tag/reference#googletag.PubAdsService.setForceSafeFrame) through the Google Publisher Tag API. But the downside is less rich media formats so less revenue.

As @patmmccann mentions it is possible to modify insert element but in practice we found this approach quite unwieldly. We did try a function to check all existing scripts/iframes at the start of a page load and then on every subsequent refresh we would remove the ones that were not necessary. Neither option is really ideal but does limit the memory usage from getting out of control.

**Comment by alnagar** [comment]
@12q1 @patmmccann 
Thanks for your insights.

I understand that we can ease the issue by work it around with safeframes or hacky solutions like monkey patching but is not this something that could be fixed within the prebidjs codebase ? Or is this not considered a bug ?

**Comment by invalidtask** [comment]
Can confirm this is still an ongoing issue. Manually removing iframes, meta tags with "origin-trial" labels, and pruning a constantly growing number of script tags, seems to slow or reduce memory usage. Any progress on addressing this issue? It's been happening for over a year now.

**Comment by invalidtask** [comment]
Bump. Still a critical problem on the latest build.

**Comment by simone-gross** [comment]
Any updates on this issue how to workaround or if it can be solved in the prebid codebase? (safeframe can not be forced because of business requirements like @12q1 already mentioned )

We also face an exhaustive memory issue on our single page application (using react) as well, caused on ad refresh as well as navigating form page to page. Exactly as already described in this issue. 
We have also applied our prebid config object with the following settings:
```
pbjs.setConfig({
        minBidCacheTTL: 0,
        eventHistoryTTL: 0,
});
```

**Comment by 12q1** [comment]
@simone-gross 

Not proud of this one but within our own script we have a function to remove third party iframes. Helps to keep the memory usage under control and also helps the garbage collection. On single page applications we execute this function when there is a navigation event to clear as much DOM pollution as possible. 

*psuedo-code example*

```
var removeAllThirdPartyIframes = () => {
    document.querySelectorAll('iframe').forEach(iframe => {
        try {
            const src = iframe.src;
            if (src && new URL(src).origin !== window.location.origin) {
                // Other blacklist conditions
                iframe.remove();
            }
        } catch (e) {
            // Silently fail if iframe.src is inaccessible (e.g., about:blank, sandboxed)
        }
    });
};
```

The downside is that you might also remove iframe embeds that you or the client wish to keep. In those cases you need to tweak the conditional logic to match your requirements (ex. allow iframes from youtube/twitter/etc).

**Comment by invalidtask** [comment]
It's sad that this problem is being largely ignored seeing as Prebid.js is one of the most widely used libraries for bidding. It's killing user experience and causing long term revenue losses.

**Comment by dgirardi** [comment]
Prebid can remove frames that it creates itself (such as user syncs). But they are left over by creatives, we can't really tell when it's OK to remove them. The best we could do is an optional feature that aggressively removes all of them as you do above.
